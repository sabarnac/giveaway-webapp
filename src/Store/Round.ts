import { observable, computed } from "mobx";
import Participant from "./Participant";
import Config from "./Config/Config";
import chunk from "lodash.chunk";
import Match from "./Match";

/**
 * Class containing the details of a round of the tournament.
 */
export default class Round {
  @observable public matches: Match[];

  private _config: Config;

  private _shuffleParticipants = (participants: Participant[]): Participant[] =>
    this._config.randomGenerator.shuffle(participants);

  private _createMatch = (participants: Participant[]): Match =>
    new Match(this._config, participants);

  private _getMatches = (participants: Participant[]): Match[] =>
    chunk(
      this._shuffleParticipants(participants),
      this._config.participantsPerMatch,
    ).map(this._createMatch);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this.matches = this._getMatches(participants);
  }

  private _getMatchParticipants = (match: Match): Participant[] =>
    match.participants;

  private _getMatchWinner = (match: Match): Participant => match.winner;

  private _getMatchLosers = (match: Match): Participant[] => match.losers;

  /**
   * Get the list of participants in the round.
   * @returns {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this.matches.map(this._getMatchParticipants).flat();
  }

  /**
   * Get the list of winners in the round.
   * @returns {Participant[]} The list of winners.
   */
  @computed public get winners(): Participant[] {
    return this.matches.map(this._getMatchWinner);
  }

  /**
   * Get the list of losers in the round.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this.matches.map(this._getMatchLosers).flat();
  }
}
