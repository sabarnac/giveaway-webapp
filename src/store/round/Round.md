Usage:

```tsx
import Match from "./store/round/match/Match";
import Participant from "./store/round/match/participant/Participant";

const round: Round = new Round(Config.instance, [
  new Participant("FOO"),
  new Participant("BAR")
]);
const roundId: Round = round.id; // "round-1"
const matches: Match[] = round.matches; // [Match("match-1"), ..., Match("match-x")]
const firstMatch: Match = round.firstMatch; // Match("match-1")
const lastMatch: Match = round.lastMatch; // Match("match-x")
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const winners: Participant[] = round.winners; // [Participant("FOO")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
```
