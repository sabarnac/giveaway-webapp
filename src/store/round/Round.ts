import { observable, computed } from "mobx";
import Participant from "./match/participant/Participant";
import Config from "../config/Config";
import chunk from "lodash.chunk";
import Match from "./match/Match";
import last from "lodash.last";
import RandomGenerator from "../config/RandomGenerator";

/**
 * Class containing the details of a round of the tournament.
 */
export default class Round {
  @observable private _id: string;
  @observable private _matches: Match[];

  private static counter: number = 1;

  private _config: Config;

  private _shuffleParticipants = (participants: Participant[]): Participant[] =>
    RandomGenerator.shuffle(participants);

  private _createMatch = (participants: Participant[]): Match =>
    new Match(participants);

  private _getMatches = (participants: Participant[]): Match[] =>
    chunk(
      this._shuffleParticipants(participants),
      this._config.participantsPerMatch
    ).map(this._createMatch);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this._id = `round-${Round.counter++}`;
    this._matches = this._getMatches(participants);
  }

  private _getMatchParticipants = (match: Match): Participant[] =>
    match.participants;

  private _getMatchWinner = (match: Match): Participant => match.winner;

  private _getMatchLosers = (match: Match): Participant[] => match.losers;

  /**
   * Get the match ID.
   * @returns {string} The unique ID of the match.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Get the list of matches in the round.
   * @returns {Match[]} The list of matches.
   */
  @computed public get matches(): Match[] {
    return this._matches;
  }

  /**
   * Get the first match of the round.
   * @returns {Match} The first match.
   */
  @computed public get firstMatch(): Match {
    return this._matches[0];
  }

  /**
   * Get the last match of the round.
   * @returns {Match} The last match.
   */
  @computed public get lastMatch(): Match {
    return <Match>last(this._matches);
  }

  /**
   * Get the list of participants in the round.
   * @returns {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this._matches.map(this._getMatchParticipants).flat();
  }

  /**
   * Get the list of winners in the round.
   * @returns {Participant[]} The list of winners.
   */
  @computed public get winners(): Participant[] {
    return this._matches.map(this._getMatchWinner);
  }

  /**
   * Get the list of losers in the round.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._matches.map(this._getMatchLosers).flat();
  }
}
