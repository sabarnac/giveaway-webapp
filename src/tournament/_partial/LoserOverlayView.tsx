import LoserOverlayView from "./LoserOverlayView.react";

export default LoserOverlayView;
