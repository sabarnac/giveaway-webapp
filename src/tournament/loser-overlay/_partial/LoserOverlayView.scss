.loser-overlay {
  border: 1px dashed #000000;
  position: relative;
  background: #ffffff;
  max-width: 80vw;
  min-width: 50vw;
  max-height: 90vh;
  overflow-y: auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 4rem 4rem 2rem 4rem;
}
