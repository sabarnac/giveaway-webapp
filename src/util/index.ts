export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
