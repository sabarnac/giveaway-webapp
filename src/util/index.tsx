import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
