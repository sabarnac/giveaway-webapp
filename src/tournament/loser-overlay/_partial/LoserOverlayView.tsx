import React, { useEffect } from "react";
import { Observer } from "mobx-react";
import "./LoserOverlayView.scss";
import classNames from "classnames";
import Participant from "../../../store/round/match/participant/Participant";
import {
  AnimationStateHookResult,
  useAnimationState,
  runOnPredicate,
} from "../../../util/index";
import LoserInfo from "./LoserInfo";

/**
 * Properties of the loser overlay React component.
 */
interface LoserOverlayViewProps {
  /** CSS class */
  className: string;
  /** The losers of a round. */
  losers: Participant[];
  /** Action to call when the overlay has finished showing the list of losers. */
  onViewComplete: () => void;
}

/**
 * React component for the loser overlay.
 */
export default (props: LoserOverlayViewProps): JSX.Element => {
  const [
    currentState,
    updateState,
  ]: AnimationStateHookResult = useAnimationState();

  useEffect(
    runOnPredicate(currentState === props.losers.length, props.onViewComplete),
  );

  return (
    <Observer>
      {() => (
        <div className={classNames(props.className)}>
          <h2>Losers</h2>
          {props.losers.map(
            (participant: Participant, index: number): JSX.Element => (
              <LoserInfo
                key={participant.name}
                className={props.className}
                show={currentState === index}
                participant={participant}
                onInfoComplete={updateState}
              />
            ),
          )}
        </div>
      )}
    </Observer>
  );
};
