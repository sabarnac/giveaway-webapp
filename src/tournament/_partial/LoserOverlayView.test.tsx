import React from "react";
import LoserOverlayView from "./LoserOverlayView";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyRound } from "../../util/test";
import Round from "../../store/round/Round";

describe("LoserOverlayView Component.", () => {
  it("Matches snapshot.", () => {
    const round: Round = createDummyRound();
    const component: ReactTestRenderer = create(
      <LoserOverlayView
        round={round}
        show={true}
        onOverlayComplete={() => {}}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
