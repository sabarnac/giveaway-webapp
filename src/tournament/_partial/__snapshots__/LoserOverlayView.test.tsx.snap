// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoserOverlayView Component. Matches snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "loser-overlay-wrapper--entering",
      "enterDone": "loser-overlay-wrapper--entered",
      "exit": "",
      "exitActive": "loser-overlay-wrapper--exiting",
      "exitDone": "loser-overlay-wrapper--exited",
    }
  }
  in={false}
  mountOnEnter={true}
  onExited={[Function]}
  timeout={500}
  unmountOnExit={true}
>
  <div
    className="loser-overlay-wrapper"
    style={
      Object {
        "transition": "opacity 500ms ease-in-out",
      }
    }
  >
    <div
      className="loser-overlay"
    >
      <h2>
        Losers
      </h2>
      <div
        classNames={
          Object {
            "enter": "",
            "enterActive": "loser-overlay__loser--entering",
            "enterDone": "loser-overlay__loser--entered",
            "exit": "",
            "exitActive": "loser-overlay__loser--exiting",
            "exitDone": "loser-overlay__loser--exited",
          }
        }
        in={false}
        mountOnEnter={true}
        onEntered={[Function]}
        onExited={[Function]}
        timeout={200}
        unmountOnExit={true}
      >
        <div
          className="loser-overlay__loser"
          style={
            Object {
              "transition": "opacity 200ms ease-in-out",
            }
          }
        >
          <div
            className="participant participant-card"
          >
            <div
              className="participant-card__avatar"
            >
              <img
                alt="foobar-2"
                src="foobar-2"
              />
            </div>
            <div
              className="participant-card__name"
            >
              <strong>
                Fozbaz 2
              </strong>
            </div>
          </div>
        </div>
      </div>
      <div
        classNames={
          Object {
            "enter": "",
            "enterActive": "loser-overlay__loser--entering",
            "enterDone": "loser-overlay__loser--entered",
            "exit": "",
            "exitActive": "loser-overlay__loser--exiting",
            "exitDone": "loser-overlay__loser--exited",
          }
        }
        in={false}
        mountOnEnter={true}
        onEntered={[Function]}
        onExited={[Function]}
        timeout={200}
        unmountOnExit={true}
      >
        <div
          className="loser-overlay__loser"
          style={
            Object {
              "transition": "opacity 200ms ease-in-out",
            }
          }
        >
          <div
            className="participant participant-card"
          >
            <div
              className="participant-card__avatar"
            >
              <img
                alt="foobar-4"
                src="foobar-4"
              />
            </div>
            <div
              className="participant-card__name"
            >
              <strong>
                Fozbaz 4
              </strong>
            </div>
          </div>
        </div>
      </div>
      <div
        classNames={
          Object {
            "enter": "",
            "enterActive": "loser-overlay__loser--entering",
            "enterDone": "loser-overlay__loser--entered",
            "exit": "",
            "exitActive": "loser-overlay__loser--exiting",
            "exitDone": "loser-overlay__loser--exited",
          }
        }
        in={false}
        mountOnEnter={true}
        onEntered={[Function]}
        onExited={[Function]}
        timeout={200}
        unmountOnExit={true}
      >
        <div
          className="loser-overlay__loser"
          style={
            Object {
              "transition": "opacity 200ms ease-in-out",
            }
          }
        >
          <div
            className="participant participant-card"
          >
            <div
              className="participant-card__avatar"
            >
              <img
                alt="foobar-6"
                src="foobar-6"
              />
            </div>
            <div
              className="participant-card__name"
            >
              <strong>
                Fozbaz 6
              </strong>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
